BOJ 12886: 돌 그룹

kdozlo opened the pull request:
## 💡 문제
[BOJ 12886: 돌 그룹](https://www.acmicpc.net/problem/12886)
<br>

## 📱 스크린샷
<img width="950" alt="image" src="https://github.com/user-attachments/assets/29bd65c0-a493-4bb9-8685-cb2b7d4bb941">
<br>

## 📝 리뷰 내용
**느린 구현 방식**
dfs 이용해서 모든 경우에 대해 조건에 만족하는 경우를 찾으려고 했습니다. 현재 문제에서는 어차피 모든 경우를 비교해서 답을 구하므로 dfs, bfs 차이가 크게 없다고 판단했습니다. 하지만 시간 초과가 났습니다.

이유는 서로 다른 숫자만 골라서 문제에서 원하는 계산을 해줘야 하는데 같은 숫자도 고려했습니다. 
또다른 이유는 방문 처리 과정에서 중복 문제가 생깁니다. 제가 생각한 방문 방식은 모든 숫자를 문자열로 변환하여 set 자료구조에 넣는 방식이였습니다. 이 경우 8, 10, 9 형태는 8109로 저장이 됩니다.  따라서 10, 8, 9도 같은 경우이지만 다른 경우로 판단합니다.

첫번째 문제 해결 방법은
두 숫자를 골랐을때 같은 숫자인지 판단하면 됩니다.

두번재 문제를 해결하는 방법은
세 숫자를 넣을 배열을 만들어 sort 하는 과정을 넣고, 모든 숫자에 대해 문자열로 변환 후 set 자료구조에 넣도록 했습니다.

이런 방식으로 구현하면 느리게 통과가 됩니다. 하지만 느린 이유는 각 순간마다 소팅과정 nlogn이 들어가기 때문입니다. 메모리 또한 많이 먹는데 각 순간마다 배열을 만들기고 하고, set 자료구조에 많은 데이터 값이 들어가기 때문입니다.

이 방식은 위험할 수도 있습니다. 예시로 33, 33, 33 숫자를 넣었다고 가정한다면 set 자료구조에는 333333으로 들어가게 될 텐데 이는 333, 3, 33으로 해석될 여지가 있습니다. 따라서 333, 3, 33가 먼저 나온다면 방문처리가 333333으로 되어 추후에 33 33 33이 되었을 때 이를 방문하지 못하게 되어 답이 나오지 않을 경우가 생기게 됩니다.

이 문제 특성상 이런 상황을 만족하는 경우가 나오지 않을 수도 있습니다. 또한 세 숫자 사이에 ,같은 표시를 통해 해결할 수 있습니다. 하지만 이 구현 방식은 너무 느립니다.

**빠른 구현 방식**
여기서 다른 두 숫자를 골라 b > a인 경우 a + a, b - a 해준다고 했습니다. 이게 무슨 뜻이냐면 고른 두 숫자를 합하면 위의 계산을 하기 전이나 후나 합이 같다는 뜻입니다. (a + b, a + a + (b - a)) 그래서 두 숫자만 알면 나머지 한 숫자는 자동으로 알 수 있습니다. 따라서 방문 체크를 visited[a][b] 형태로 해주면 됩니다. 이렇게 한다면 메모리를 많이 잡아 먹지도 않고, 각 재귀마다 소팅 과정도 생략되기 때문에 시간이 빨라집니다.

**느낀점**
문제의 조건에서 숨은 뜻을 찾는게 어렵네요. 못찾아서 분합니다. 

**추가 구현**
리뷰 내용 반영해서 코드 수정 했습니다. 속도가 196ms -> 169ms 까지 내렸네요. @tddudwns1  감사합니다.
1. dfs 내 포문 범위 변경
2. 곱하기 2 연산 비트 연산자로 변경
3. 방문 체크 추가
    a. 전 코드 : a,b/b,a
    b. 리펙토링 : a,b/b,a/a,c/c,a/c,b/b,c

추가로 visited 배열 if문 안에 넣는건 해보니 더 느려지더라고요. 아래 리뷰 답글 참고해 보세요.

kdozlo committed:
BOJ 12886: 돌 그룹

- Set, Arrays.sort()를 이용한 느린 버전

kdozlo committed:
BOJ 12886: 돌 그룹

tddudwns1 commented on kimdaeyeong/src/baekjoon/Boj12886.java:
```diff
@@ -0,0 +1,64 @@
+
+        for(int i = 0; i < 3; i++) {
+            for(int j = i + 1; j < 3; j++) {
+                if(array[i] != array[j]) {
```
이 조건문은 필요 없을 것 같습니다! 하하~

kdozlo replied:
이미 필요없는 연산 인정합니다~!

kdozlo replied:
아아니다 이거 필요해요. 배열에 다른 인덱스를 선택했다고 해도 값은 같을 가능성이 있습니다.

tddudwns1 replied:
어 맞네요 ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ

tddudwns1 replied:
i != j 랑 헷갈려버리기~

kdozlo replied:
제 코드에 혼란을~~ 무섭네요

tddudwns1 commented on kimdaeyeong/src/baekjoon/Boj12886.java:
```diff
@@ -0,0 +1,64 @@
+        boolean[][] visited = new boolean[1501][1501];
+        int sum = A + B + C;
+        int answer = 0;
+        visited[A][B] = true;
+        visited[B][A] = true;
```
AC CA BC CB도 확인 하면 좋을 것 같습니다! 하하~

kdozlo replied:
이건 진짜 좋은데요. 속도가 굉장히 업될 느낌 우와우

tddudwns1 commented on kimdaeyeong/src/baekjoon/Boj12886.java:
```diff
@@ -0,0 +1,64 @@
+        int answer = 0;
+        visited[A][B] = true;
+        visited[B][A] = true;
+        if(sum % 3 == 0 && dfs(sum, A, B, visited)) {
+            answer = 1;
+        }
```
위의 visited 배열 생성과 방문 체크를 이 조건문 내부에 넣으면 좋을 것 같습니다! 하하~

kdozlo replied:
visited 배열 생성을 조건문 내부에서만 생성되게 하기 위한 의도인거죠? 그 외 다른 의도가 있나요?

tddudwns1 replied:
정답! 다른 의도는 없슴다~ 하하~

kdozlo replied:
if(sum % 3 == 0) {
            boolean[][] visited = new boolean[1501][1501];
            visited[A][B] = true;
            visited[B][A] = true;

            if(dfs(sum, A, B, visited))
                answer = 1;
        }
        
       이렇게 바꿨는데 더 느리더라구요...

tddudwns1 replied:
제가 테스트해봤는데 더 느려지진 않았습니다
그럴 수도 없구요

아마 백준 서버차이 같슴당나귀~

kdozlo replied:
서버 차이라면 필요한 경우만 visited 배열 생성하는게 좋아 보이긴 하네요

tddudwns1 replied:
975

tddudwns1 commented on kimdaeyeong/src/baekjoon/Boj12886.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    public static boolean dfs(int sum, int a, int b, boolean[][] visited) {
+        int c = sum - a - b;
```
이제보니까 sum도 넘기고 a도 넘기고 b도 넘기는데 sum의 용도가 그냥 c를 구하는거면 c를 진작 넘기면 안될까요?

kdozlo replied:
이건 a,b,c를 배열에 넣고 소팅을 해요. 그 후 두개를 골라서 다음 재귀로 넣는데 안고른 하나도 넣어야 하죠. 예를 들어 array[1], array[2]를 골랐다면 나머지 array[0]를 넣어야 하고, array[0], array[1]을 고른다면 array[2]를 골라야 해요. 근데 이렇게 알기 위해 추가 연산을 해야겠죠? 여러가지 방법이 있겠지만 제일 편한건 합을 구해서 고른 두개를 뺀 c를 구하는거 일겁니다. 
자 그럼 a, b, c를 넘긴다 하더라도 결국 sum은 또 다시 구해야 하는 상황이 오는거죠. 이럴거면 처음부터 sum을 고르는게 좋습니다.

tddudwns1 replied:
자 확실하진 않지만 지금 생각난 대로 말씀 드리겠습니다

먼저 dfs의 파라미터로 int 3개가 아닌 int배열을 받습니다

그리고 하단의 na, nb 값을 그 배열의 위치(for문을 통해 i와 j를 알기때문)에 넣습니다
그러고 그대로 넘기면 될 것 같습니다
어떤가요?
![image](https://github.com/user-attachments/assets/279f0f0c-5351-4acd-ad4a-5191ecae5ef5)

tddudwns1 replied:
![image](https://github.com/user-attachments/assets/881edc73-0671-42bc-8cfa-23ca3e0f49de)
성공~

kdozlo replied:
이건 궁금한게 배열으로 넘겨서 속도가 빨라긴건지 범위를 제한해서 빨리진건지 궁금하네요

kdozlo replied:
<img width="485" alt="image" src="https://github.com/user-attachments/assets/093655a0-2f05-43a0-ba78-8bbd72e7e4db">
이것도 비교해 봤는데 큰 차이는 없네요 속도가 높아진건 아무래도 범위를 축소시켜서 그런것 같아요

tddudwns1 replied:
아 근가요? 근데 사실 저도 배열 복사하면서 이게 맞나 했습니다
아님 말고~ 하하~

tddudwns1 commented on kimdaeyeong/src/baekjoon/Boj12886.java:
```diff
@@ -0,0 +1,64 @@
+        int[] array = {a, b, c};
+        boolean checked = false;
+
+        for(int i = 0; i < 3; i++) {
```
i = 0; i < 2로 하면 더 좋을 것 같습니다! 하하~

kdozlo replied:
맞네요 2로 갈 필요도 없네요.. 코드가 디테일이 좀 없었네요 리펙토링과 리뷰가 이렇게나 중요합니다. 좋은 의견 감사합니다

tddudwns1 commented on kimdaeyeong/src/baekjoon/Boj12886.java:
```diff
@@ -0,0 +1,64 @@
+        for(int i = 0; i < 3; i++) {
+            for(int j = i + 1; j < 3; j++) {
+                if(array[i] != array[j]) {
+                    int na = array[i] > array[j] ? array[i] - array[j] : array[i] * 2;
+                    int nb = array[i] < array[j] ? array[j] - array[i] : array[j] * 2;
```
*2 대신 <<1은 어떨까요? 하하~

kdozlo replied:
고급 기술 비트 연산자! 적용해볼께요


섬세한 리뷰 감사합니다. 속도가 많이 오르겠네요

kdozlo committed:
BOJ 12886: 돌 그룹

- dfs 내 포문 범위 변경
- 곱하기 2 연산 비트 연산자로 변경
- 방문 체크 추가
  - 전 코드 : a,b/b,a
  - 리펙토링 : a,b/b,a/a,c/c,a/c,b/b,c

limsubinn commented:
리뷰권을 빼앗겼다
이미 위에서 토론의 장을 펼치셔서 제가 더 할 말은 없는 것 같습니다

항상 드는 생각인데 문제를 꼼꼼하게 안 읽어서 하는 실수가 많으신 것 같아요. 
물론 저도 그런 적 많지만,, 이 부분 좀 신경 써야 할 것 같습니다.
고생하셨어요.

kdozlo commented:
> 리뷰권을 빼앗겼다 이미 위에서 토론의 장을 펼치셔서 제가 더 할 말은 없는 것 같습니다
> 
> 항상 드는 생각인데 문제를 꼼꼼하게 안 읽어서 하는 실수가 많으신 것 같아요. 물론 저도 그런 적 많지만,, 이 부분 좀 신경 써야 할 것 같습니다. 고생하셨어요.

저도 스터디 하면서 확실히 느끼는게 문제를 꼼꼼히 안 읽는 버릇이 있었습니다. 고쳐나가기 위해 요즘 메모하며 풉니다 하하
